File: 2021/09/02/repeated_substring.py
"""
https://leetcode.com/problems/repeated-substring-pattern/
Author: Andrew Jarombek
Date: 9/2/2021
"""


def repeated_substring_pattern(s: str) -> bool:
    for i in range(1, (len(s) // 2) + 1):
        if s[i:] + s[:i] == s:
            return True

    return False


def repeated_substring_pattern_v2(s: str) -> bool:
    return s in (s + s)[1:-1]


if __name__ == '__main__':
    assert repeated_substring_pattern("abab")
    assert not repeated_substring_pattern("aba")
    assert repeated_substring_pattern_v2("abab")
    assert not repeated_substring_pattern_v2("aba")

File: 2023/06/27/go/repeated_substring.go
/**
 * https://leetcode.com/problems/repeated-substring-pattern/
 * Author: Andrew Jarombek
 * Date: 6/27/2023
 */

package _go

func repeatedSubstringPattern(s string) bool {
	for i := 1; i <= len(s)/2; i++ {
		if len(s)%i == 0 {
			sub := s[0:i]
			repeated := true

			for j := i; j < len(s); j += i {
				if sub != s[j:j+i] {
					repeated = false
					break
				}
			}

			if repeated {
				return true
			}
		}
	}

	return false
}
